internal/storage/metricstore/elasticsearch/reader.go

```diff
@@ -6,47 +6,371 @@ package elasticsearch
+			})
+		}
+
+		metric.MetricPoints = processedPoints
+	}
+
+	return mf
+}
+
+func getCallRateBucketsToPoints(buckets []*elastic.AggregationBucketHistogramItem) []*Pair {
+	var points []*Pair
+
+	for _, bucket := range buckets {
+		aggMap, ok := bucket.Aggregations.CumulativeSum(culmuAggName)
+		if !ok {
+			return nil
+		}
+		value := math.NaN()
+		if aggMap != nil && aggMap.Value != nil {
+			value = *aggMap.Value
+		}
+		points = append(points, &Pair{
+			TimeStamp: int64(bucket.Key),
+			Value:     value,
+		})
+	}
+	return points
+}
+
+// buildCallRateAggregations constructs the GetCallRate aggregations.
+func (MetricsReader) buildCallRateAggregations(params metricstore.BaseQueryParameters, timeRange TimeRange) elastic.Aggregation {
+	fixedIntervalString := strconv.FormatInt(params.Step.Milliseconds(), 10) + "ms"
+	dateHistoAgg := elastic.NewDateHistogramAggregation().
+		Field("startTimeMillis").
+		FixedInterval(fixedIntervalString).
+		MinDocCount(0).
+		ExtendedBounds(timeRange.extendedStartTimeMillis, timeRange.endTimeMillis)
+
+	cumulativeSumAgg := elastic.NewCumulativeSumAggregation().BucketsPath("_count")
+
+	// Corresponding AGG ES query:
+	// "aggs": {
+	//	"results_buckets": {
+	//		"date_histogram": {
+	//			"field": "startTimeMillis",
+	//				"fixed_interval": "60s",
+	//				"min_doc_count": 0,
+	//				"extended_bounds": {
+	//				"min": "now-lookback-5m",
+	//				"max": "now"
+	//			}
+	//		},
+	//		"aggs": {
+	//			"cumulative_requests": {
+	//				"cumulative_sum": {
+	//					"buckets_path": "_count"
+	//				}
+	//			}
+	//
+
+	dateHistoAgg = dateHistoAgg.
+		SubAggregation(culmuAggName, cumulativeSumAgg)
+
+	if params.GroupByOperation {
+		operationsAgg := elastic.NewTermsAggregation().
+			Field("operationName").
+			Size(10).
+			SubAggregation("date_histogram", dateHistoAgg) // Nest the dateHistoAgg inside operationsAgg
+		return operationsAgg
+	}
+
+	return dateHistoAgg
+}
+
+// executeSearch performs the Elasticsearch search.
+func (r MetricsReader) executeSearch(ctx context.Context, p MetricsQueryParams) (*metrics.MetricFamily, error) {
+	if p.GroupByOperation {
+		p.metricName = strings.Replace(p.metricName, "service", "service_operation", 1)
+		p.metricDesc += " & operation"
+	}
+
+	span := r.queryLogger.TraceQuery(ctx, p.metricName)
+	defer span.End()
+
+	indexName := r.cfg.Indices.IndexPrefix.Apply("jaeger-span-*")
+	searchResult, err := r.client.Search(indexName).
+		Query(&p.boolQuery).
+		Size(0). // Set Size to 0 to return only aggregation results, excluding individual search hits
+		Aggregation(aggName, p.aggQuery).
+		Do(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed executing metrics query: %w", err)
+		r.queryLogger.LogErrorToSpan(span, err)
+		return nil, err
+	}
+
+	r.queryLogger.LogAndTraceResult(span, searchResult)
+
+	rawResult, err := ToDomainMetricsFamily(p, searchResult)
+	if err != nil {
+		return nil, err
+	}
+
+	processedResult := p.processMetricsFunc(rawResult, p.BaseQueryParameters)
```

Comment:
there is no need for this processMetricsFunc call to happen inside executeSearch, that's exactly the indirection I did not like in the comment above. executeSearch just returns the result, and the caller can do additional processing without creating any lambdas,

Reply:
I am going to merge this as is for now, but please address in the follow-up PR.